staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -453,19 +453,21 @@ func (s sortableStoreElements) Swap(i, j int) {
 // WaitUntilFreshAndList returns list of pointers to `storeElement` objects along
 // with their ResourceVersion and the name of the index, if any, that was used.
 func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, key string, matchValues []storage.MatchValue) ([]interface{}, uint64, string, error) {
-	items, rv, index, err := w.waitUntilFreshAndListItems(ctx, resourceVersion, key, matchValues)
+	result, rv, index, filteredAndOrdered, err := w.waitUntilFreshAndListItems(ctx, resourceVersion, key, matchValues)
 	if err != nil {
 		return nil, 0, "", err
 	}
-	result, err := filterPrefix(key, items)
-	if err != nil {
-		return nil, 0, "", err
+	if !filteredAndOrdered {
+		result, err = filterPrefix(key, result)
+		if err != nil {
+			return nil, 0, "", err
+		}
+		sort.Sort(sortableStoreElements(result))
 	}
-	sort.Sort(sortableStoreElements(result))
 	return result, rv, index, nil
 }
 
-func (w *watchCache) waitUntilFreshAndListItems(ctx context.Context, resourceVersion uint64, key string, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, err error) {
+func (w *watchCache) waitUntilFreshAndListItems(ctx context.Context, resourceVersion uint64, key string, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, filteredAndOrdered bool, err error) {
```

Comment:
Please explain that `filteredAndOrdered` refers only to filtering by key prefix, not about selectors.

Reply:
Done